src/entities/swap/swaps/v3/index.ts

```diff
@@ -0,0 +1,525 @@
+                        DEFAULT_USERDATA,
+                    ],
+                    { blockNumber: block },
+                );
+            return {
+                swapKind: SwapKind.GivenOut,
+                expectedAmountIn: TokenAmount.fromRawAmount(
+                    this.inputAmount.token,
+                    result,
+                ),
+            };
+        }
+        throw Error('Unsupported V3 Query');
+    }
+
+    private async queryBatchSwap(
+        client: PublicClient,
+        block?: bigint,
+    ): Promise<ExactInQueryOutput | ExactOutQueryOutput> {
+        // Note - batchSwaps are made via the Batch Router
+        const routerContract = getContract({
+            address: BALANCER_BATCH_ROUTER[this.chainId],
+            abi: balancerBatchRouterAbi,
+            client,
+        });
+        /*
+        In V3 all paths must have individual limits set using minAmountOut/maxAmountIn. 
+        pathAmountsOut/In returned by query can be used along with slippage to set these correctly.
+        */
+        if (this.swapKind === SwapKind.GivenIn) {
+            const swapsWithLimits = (this.swaps as SwapPathExactAmountIn[]).map(
+                (s) => {
+                    return {
+                        ...s,
+                        minAmountOut: 0n, // we default to 0 for query
+                    };
+                },
+            );
+            const { result } = await routerContract.simulate.querySwapExactIn(
+                [swapsWithLimits, DEFAULT_USERDATA],
+                { blockNumber: block },
+            );
+
+            if (result[1].length !== 1)
+                throw Error(
+                    'Swaps only support paths with matching tokenIn>tokenOut',
+                );
+
+            return {
+                swapKind: SwapKind.GivenIn,
+                expectedAmountOut: TokenAmount.fromRawAmount(
+                    this.outputAmount.token,
+                    result[2][0],
+                ),
+                pathAmounts: result[0] as bigint[],
+            };
+        }
+
+        const swapsWithLimits = (this.swaps as SwapPathExactAmountOut[]).map(
+            (s) => {
+                return {
+                    ...s,
+                    maxAmountIn: 0n, // we default to 0 for query
+                };
+            },
+        );
+        const { result } = await routerContract.simulate.querySwapExactOut(
+            [swapsWithLimits, DEFAULT_USERDATA],
+            { blockNumber: block },
+        );
+
+        if (result[1].length !== 1)
+            throw Error(
+                'Swaps only support paths with matching tokenIn>tokenOut',
+            );
+
+        return {
+            swapKind: SwapKind.GivenOut,
+            expectedAmountIn: TokenAmount.fromRawAmount(
+                this.inputAmount.token,
+                result[2][0],
+            ),
+            pathAmounts: result[0] as bigint[],
+        };
+    }
+
+    public queryCallData(): string {
```

Comment:
This method feels a bit redundant with `query`. Can we build query data and then reuse it within query to avoid code duplication? Or is different in any way?

Reply:
And maybe rename to `buildQuery` to be consistent with `buildCall`?

Reply:
`queryCallData` is a function from the original Swaps implementation. I originally thought it was a bit redundant too but I think the goal is to provide the encoded data so a user can use it to call efficiently whenever and however they want without have to re-encode, etc.
The only way I can see to use the call data in `query` is to use [call](https://viem.sh/docs/actions/public/call) but that returns encoded data which we would then need to decode. `simulate` seems like the better choice here.
I've added a helper, `getSwapsWithDefaultLimits()` that can be used in both to reuse as much common code as possible.

Reply:
Got it 👍 
I agree that being able to use simulate is better than reusing a bit more code.